Added beginnings of a developer quick-start guide

pdaoust opened the pull request:
Hey, dev kit folks. @dhtnetwork and I are working on a small set of resources to quickly acquaint development teams with Holochain. The purpose is to help them quickly understand, at a very high level:

* why we're building Holochain
* why they might want to use it, including an entire document on why it's way cooler than blockchain
* what are the building blocks at their disposal
* how they might go about designing their app's architecture

This package is meant to serve a few different purposes: one, to help developers quickly grasp what the heck this thing actually is; two, orient people coming to meetups and hackathons; three, to reduce time spent on the basics during our discovery calls with dev teams.

Some of you have already given feedback on both of these documents (thank you!).

**Anyhow**, where was I... oh yes! We thought this repo would be a good place for these documents to live -- we'd love to see your reviews and edits. So I humbly submit this for your scrutiny and hopeful acceptance into `master`.

On my to-do list:

* Reduce word count
* Possibly restructure the longer introduction
* Come up with some better document titles
* Replace lots of the text with graphics
* Write a third document, on Holochain vs blockchain (the master doc has been edited; just need to select content for a shorter doc)

pdaoust committed:
Add brief introduction

pdaoust committed:
Added longer introduction

pdaoust committed:
Improved language around risks to service operators

pdaoust committed:
Added document purpose

d3sbar commented on holochain-introduction-for-developers/holochain-brief-introduction.md:
```diff
@@ -0,0 +1,96 @@
+
+### Resilience through interconnection
+
+Everyone is part of a peer-to-peer network that replicates public data, audits it, and spreads warnings of bad actors. Third-party validators are chosen at random to thwart collusion.
```
"that replicates public data"... what about the ability to distribute private data?? I have a membrane to mark my DHT submission as private - but want the resilience (redundancy) for my data - is that to be alluded here or another feature?

pdaoust replied:
@d3sbar could you go into the use case a little more? Private data, by nature, is not distributed. I can see a use case for 'hidden in plain sight' data -- that is, data that's encrypted before it's sent to the DHT. I feel like it's a decent pattern for data that's meant for a few eyes only, less appropriate for storing completely private data (entry types that would normally be marked 'private').

Thoughts:

* Arthur recommends against this. His big concern is that all cryptography eventually gets broken, given enough time, and do you really want that data living in a DHT forever? I personally think this is more of a theoretical concern than a pragmatic one.
* If you're encrypting using a symmetric key (which you should -- more quantum resistant and much faster than public key encryption) you typically distribute the symmetric key to intended recipients by first encrypting it with their public key. This is how most public-key encryption _actually_ works behind the scenes.
* Regardless of whether you encrypt using a symmetric key or your recipients' public keys, you will never be able to revoke access to that data once you've put it out there. And this reflects reality anyway: the first time your recipient opens the encrypted data, they can save it in unencrypted form on their hard drive.
* Encrypted data can't be validated by peers because they can't decrypt it. The validation function would have to be something like `return true`. In this scenario, validating peers are reduced to just ignorant deputies. This probably isn't a problem, but it depends on your app's needs.

If the big concern is resilience for private source chain entries, I believe there will be an app for backing up private source chain data and restoring it to your device.

pdaoust replied:
@d3sbar do you feel like this sentence is ambiguous and implies that all data is public (a la blockchain)?

d3sbar commented on holochain-introduction-for-developers/holochain-longer-introduction.md:
```diff
@@ -0,0 +1,190 @@
+# Missing subjects
+
+* Core services and apps -- DPKI, Vault (touched on HCHC briefly under [delivering section](#Delivering-your-application-to-the-world)). I've heard that there are also core 'template apps' (for lack of a better phrase). I don't actually know what these are, but they seem to be things like basic DNAs for messaging, commenting, etc.
+* Details of DPKI and what it's supposed to do -- this will be mostly uninteresting from people coming from the server-based world (except that they won't have to learn how to store passwords, which is cool). But it will be an amazing epiphany for people coming from the blockchain world.
```
Actually - the implementation is different than a server app. There is no centralized database or directory - so therefore a unique design pattern exists along with personas, membranes, etc. That needs to be a separate document most likely and therefore a link here and summarizing sentence referencing the link.

pdaoust replied:
For sure. This is trying to be the briefest introduction to all the puzzle pieces, so linking is the right thing here. I'm starting to feel like DPKI is a big selling point for both server-based developers (no more data leaks!) and blockchain developers (no lost keys!)

pdaoust committed:
Roll some missing subjects into article

I got rid of the 'missing subjects' section and made brief mention of:

* DPKI and Vault (under the 'User identities' section)
* third-party generic DNA templates, including ones built by Holo (blurb about remixing existing DNAs into your app)
* removed bit about onboarding process and admin GUI

pdaoust committed:
Beginning of a list of things to think about

Thoughts for developers to ponder. Seems a bit long, yet also a bit short. As we write articles digging into each of these issues, we can shorten the bullets and link out to said articles.

pdaoust committed:
many small text changes

for readability and clarity

pdaoust committed:
Clarify the audience

pdaoust committed:
clarify audience in longer intro

pdaoust committed:
Lots of simplification, improvement, etc

pdaoust committed:
even more editing for brevity

pdaoust committed:
tiny edit: remove 'app store' mention

It was kinda redundant with 'software distribution platform', which comes right afterwards.

pdaoust committed:
More content improvements

zippy commented on holochain-introduction-for-developers/holochain-brief-introduction.md:
```diff
@@ -0,0 +1,75 @@
+
+## The problem
+
+The Internet is decentralized by design. Many services, such as the web and email, have no central points. But in order to move through the Internet confidently, we need **trust.** How can you know I am who I say I am, and how can you be sure that I'm acting with integrity?
```
Um, only parts of the internet, especially routing.  Other parts are quite centralized.  DNS is centralized (how many top-level dns servers are there, 12?) IP address allocation is centralized.  And it's architecture yields centralizing pareto effects as you note.  The problem is that the Internet isn't _fully_ decentralized.   Maybe the initial sentence wouldn't have got me if it were more like:  "We often think that the internet is decentralized by design... "

pdaoust replied:
okay, point noted. I guess that just saying 'The Internet' is a bit vague, because while the routing layer is decentralisation-friendly, the systems that support it are emphatically not. I'll chew on this.

pdaoust replied:
changed some of the language here; the new diff will show it.

zippy commented on holochain-introduction-for-developers/holochain-brief-introduction.md:
```diff
@@ -0,0 +1,75 @@
+
+## Holochain: fully sovereign, distributed, trustable coordination
+
+Holochain gives you the tools to create applications that enable direct, trustable interactions among sovereign peers. It does this without needing the central coordination of a server or a group of miners. We are building:
```
```suggestion
Holochain gives you the tools to create applications that enable direct, trustable interactions among mutually sovereign peers. It does this without needing the central coordination of a server or a group of miners. We are building:
```

zippy commented on holochain-introduction-for-developers/holochain-brief-introduction.md:
```diff
@@ -0,0 +1,75 @@
+* A set of **core apps and utilities**, including an identity manager, a private data vault, and a software distribution platform.
+* A **distributed hosting ecosystem** called Holo Hosting which bridges Holochain to the web so people can use your apps as if they were traditional, server-based apps.
+
+### Uncompromisable sovereignty
```
I hope for the inclusion of the term "mutual" where ever we use sovereign, because I wish for us to appeal not only to individual agency, but deep cooperation in the context of that agency.  Sovereignty by itself is just the power to say #@!$#*%&# off!  But when that is landed in the context of mutuality something more powerful and transcendent happens.   Remember Holochain is BOTH agent centric and socio-centric, in that we are all validating the rules together, and that social act is what creates the power of sovereignty.  I'm not saying you have to change a word.  I'm just trying to point to what to me is a crucial part of the system.

pdaoust replied:
Yes, this represents a bit of a struggle for me. On the one hand, I know it's our job to shift the Overton window, but at the same time I don't want to use too much Holochain-ese, know what I mean? I don't want to lose the fact that this is about _interdependence_, not _independence_, but I'm already struggling to be brief :D

zippy commented on holochain-introduction-for-developers/holochain-brief-introduction.md:
```diff
@@ -0,0 +1,75 @@
+
+### Versatile organizational forms
+
+Holochain doesn't impose any economic, governance, licensing, or business models on your project. You're free to use Holochain's tools to build software for a commons, a non-profit, a startup, or a huge corporation.
```
This isn't quite true.  We are working on the cryptographic automony licence, which does create some limitations.  See https://medium.com/holochain/understanding-the-cryptographic-autonomy-license-172ac920966d

pdaoust replied:
oh yeah, I guess it does. After reading everything about the CAL (a few times over), I feel like I'm 90% certain about my interpretation... tl;dr is:

* it's not viral like the GPL
* you can't control someone else's keys

Is that about right @zippy?

pdaoust committed:
Improve language re: sovereignty and cooperation

* Removed most uses of the word 'sovereign' ('empowered', 'self-determination)
* Coupled remaining use with 'mutually'
* Added new 'cooperation first' section

Trying to remove hints of "screw you" and reinforce the "social coordination" message.

pdaoust commented on holochain-introduction-for-developers/holochain-brief-introduction.md:
```diff
@@ -0,0 +1,79 @@
+
+Blockchain opened our imaginations to the possibility of trustable, decentralized coordination with untrusted peers, but suffers from efficiency problems that are inherent to its design. The next wave of decentralized ledgers improve performance and resilience but carry over some of blockchain's assumptions. They hint at the full potential of this technology but stop before they reach the finish  ...
+
+## Holochain: empowered, distributed, trustable cooperation
```
@zippy I changed a bunch of the language around sovereignty/mutuality, tried to remove the suggestion that this is for lone wolves only. Here I chose "empowered" and added "mutually" before "sovereign" in the following paragraph. Would love your feedback. Read on for more changes...

pdaoust commented on holochain-introduction-for-developers/holochain-brief-introduction.md:
```diff
@@ -0,0 +1,79 @@
+* A set of **core apps and utilities**, including an identity manager, a private data vault, and a software distribution platform.
+* A **distributed hosting ecosystem** called Holo Hosting which bridges Holochain to the web so people can use your apps as if they were traditional, server-based apps.
+
+### Self-determination
```
@zippy in this section I'm trying to focus specifically on the agency of the agent in language that's easy to understand. 'Agency' is a good word but I feel like its common meaning isn't clear enough for use here. Does "Self-determination" feel good enough?

You'll note that I also added a section just after this that shows Holochain is not meant for isolated individuals; it's meant for social interaction more than anything. Do you think that these two together is adequate for this short sales pitch? I think that, to do real justice to the concept of mutual sovereignty, I'd have to expand this into a full treatise on voice/exit, shared agreements, power dynamics, etc. Not the place for it ;)

pdaoust committed:
Fix issues with 'decentralised internet' claim

@zippy made a good point: there's nothing terribly decentralised about any part of the Internet, except the routing protocol and the assumptions built into layer 7 protocols like SMTP and Usenet. (and even those are subject to Pareto effects...)

Not 100% sure about these changes; I use the word 'trust' too many times, and I'm still not happy with the "How do you know I'm telling you the truth?" line... feels like it doesn't give enough context about what it's trying to say.
